cpp/src/components/weakly_connected_components.cu

```diff
@@ -0,0 +1,638 @@
+
+#include <thrust/copy.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/optional.h>
+#include <thrust/shuffle.h>
+#include <thrust/tuple.h>
+
+#include <limits>
+#include <type_traits>
+
+namespace cugraph {
+namespace experimental {
+
+namespace {
+
+// FIXME: this function (after modification) may be useful for SSSP with the near-far method to
+// determine the near-far threshold.
+// add new roots till the sum of the degrees first becomes no smaller than degree_sum_threshold and
+// returns a triplet of (new roots, number of scanned candidates, sum of the degrees of the new
+// roots)
+template <typename GraphViewType>
+std::tuple<rmm::device_uvector<typename GraphViewType::vertex_type>,
+           typename GraphViewType::vertex_type,
+           typename GraphViewType::edge_type>
+accumulate_new_roots(raft::handle_t const &handle,
+                     vertex_partition_device_t<GraphViewType> vertex_partition,
+                     typename GraphViewType::vertex_type const *components,
+                     typename GraphViewType::edge_type const *degrees,
+                     typename GraphViewType::vertex_type const *candidate_first,
+                     typename GraphViewType::vertex_type const *candidate_last,
+                     typename GraphViewType::vertex_type max_new_roots,
+                     typename GraphViewType::edge_type degree_sum_threshold)
+{
+  using vertex_t = typename GraphViewType::vertex_type;
+  using edge_t   = typename GraphViewType::edge_type;
+
+  // FIXME: tuning parameter (time to scan max_scan_size elements should not take significantly
+  // longer than scanning a single element)
+  vertex_t max_scan_size =
+    static_cast<vertex_t>(handle.get_device_properties().multiProcessorCount) * vertex_t{1024};
+
+  rmm::device_uvector<vertex_t> new_roots(0, handle.get_stream_view());
+  vertex_t num_scanned{0};
+  edge_t degree_sum{0};
+  while (candidate_first + num_scanned < candidate_last) {
+    auto scan_size = std::min(
+      max_scan_size,
+      static_cast<vertex_t>(thrust::distance(candidate_first + num_scanned, candidate_last)));
+
+    auto num_new_roots = static_cast<vertex_t>(new_roots.size());
+    new_roots.resize(num_new_roots + scan_size, handle.get_stream_view());
+    rmm::device_uvector<vertex_t> scan_counts(0, handle.get_stream_view());
+    {
+      rmm::device_uvector<vertex_t> indices(scan_size, handle.get_stream_view());
+      auto input_pair_first = thrust::make_zip_iterator(thrust::make_tuple(
+                                candidate_first, thrust::make_counting_iterator(vertex_t{0}))) +
+                              num_scanned;
+      auto output_pair_first = thrust::make_zip_iterator(
+        thrust::make_tuple(new_roots.begin() + num_new_roots, indices.begin()));
+      new_roots.resize(
```

Comment:
This resize is a down-sizing so does not involve memory reallocation or copying, but your comment is valid to the resize in https://github.com/rapidsai/cugraph/pull/1604/files/8d2a5c4ec2b3bf74386886c2b65a205054fb2d1b#diff-eea086cb1f3b4fd23ff600a41edce68425d996f05e3b80afdac36c7e20b273e8R84

I updated the code to potentially up-sizing new_roots multiple times (at the expense of initially allocating max_new_roots elements... this will likely to be over-allocating in most cases, but I think this will be OK as this will unlikely to be a memory footprint bottleneck).

Let me know if you have further concerns.